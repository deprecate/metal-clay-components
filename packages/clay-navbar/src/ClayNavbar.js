import 'clay-link';
import {Config} from 'metal-state';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';

import templates from './ClayNavbar.soy.js';

/**
 * Metal Clay Navbar component.
 */
class ClayNavbar extends Component {
	/**
	 * @inheritDoc
	 */
	attached() {
		this.addListener('transitionend', this.handleTransitionEnd_, true);
	}

	/**
	 * Check the click and set transition true.
	 * @param {!Event} event
	 * @private
	 */
	handleClickToggler_() {
		if (!this.isTransitioning_) {
			this.isTransitioning_ = true;
		}
	}

	/**
	 * Handle css transition end.
	 * @param {!Event} event
	 * @private
	 */
	handleTransitionEnd_(event) {
		const element = this.element.querySelector('.navbar-collapse');
		if (
			element == event.target &&
			this.isTransitioning_ &&
			!this.visible_
		) {
			this.visible_ = true;
			this.isTransitioning_ = false;
		} else if (element == event.target) {
			this.visible_ = false;
			this.isTransitioning_ = false;
		}
	}

	/**
	 * @inheritDoc
	 */
	syncIsTransitioning_() {
		const elementCollapse = this.element.querySelector('.navbar-collapse');

		if (this.isTransitioning_ && !this.visible_) {
			const heightCollapse = elementCollapse.querySelector('.container')
				.clientHeight;

			elementCollapse.setAttribute(
				'style',
				`height: ${heightCollapse}px`
			);
		} else if (this.isTransitioning_ && this.visible_) {
			elementCollapse.setAttribute('style', 'height: 0');
			elementCollapse.removeAttribute('style');
		}
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayNavbar.STATE = {
	/**
	 * The display style attribute.
	 * @instance
	 * @memberof ClayNavbar
	 * @type {?boolean}
	 * @default false
	 */
	inverted: Config.bool().value(false),

	/**
	 * The toggle animation.
	 * @instance
	 * @memberof ClayNavbar
	 * @type {?boolean}
	 * @default false
	 * @private
	 */
	isTransitioning_: Config.bool()
		.value(false)
		.internal(),

	/**
	 * List of items to show in the Navbar.
	 * @instance
	 * @memberof ClayNavbar
	 * @type {!Array}
	 * @default undefined
	 */
	items: Config.arrayOf(
		Config.shapeOf({
			active: Config.bool().value(false),
			title: Config.string().required(),
			url: Config.string().required(),
		})
	).required(),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayNavbar
	 * @type {!string}
	 * @default undefined
	 */
	spritemap: Config.string().required(),

	/**
	 * Navmenus visible in mobile when click in the button.
	 * @instance
	 * @memberof ClayNavbar
	 * @type {?bool}
	 * @default false
	 * @private
	 */
	visible_: Config.bool()
		.value(false)
		.internal(),
};

defineWebComponent('clay-navbar', ClayNavbar);

Soy.register(ClayNavbar, templates);

export {ClayNavbar};
export default ClayNavbar;
