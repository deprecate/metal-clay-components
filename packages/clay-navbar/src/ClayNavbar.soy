{namespace ClayNavbar}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param? inverted: bool}
	{@param? pages: list<?>}

	{let $attributes kind="attributes"}
		class="navbar navbar-expand-md
			{if $inverted}
				{sp}navbar-dark navbar-dark-bg
			{/if}
		"
	{/let}

	<nav {$attributes}>
		<div aria-expanded="false" class="collapse navbar-collapse" id="navbar-collapse-0a">
			{call .navmenu}
				{param pages: $pages /}
			{/call}
		</div>
	</nav>
{/template}

{template .navitem}
	{@param? active: bool}
	{@param? title: string}
	{@param? url: string}

	{let $attributes kind="attributes"}
        {if $active}
            aria-label="Current"
        {/if}

		class="nav-link
			{if $active}
				{sp}active
			{/if}
		"

		href="{$url}"
	{/let}

	<li class="nav-item">
		<a {$attributes}>
			<span class="navbar-text-truncate">{$title}</span>
		</a>
	</li>
{/template}

{template .navmenu}
	{@param? pages: list<?>}

	{if $pages}
		<ul class="navbar-nav">
			{foreach $page in $pages}
				{call .navitem}
					{param active: $page.active /}
					{param title: $page.title /}
					{param url: $page.url /}
				{/call}
			{/foreach}
		</ul>
	{/if}	
{/template}
