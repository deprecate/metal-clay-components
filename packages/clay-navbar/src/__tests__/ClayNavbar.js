'use strict';

import ClayNavbar from '../ClayNavbar';

let navbar;

describe('ClayNavbar', function() {
  afterEach(() => {
    if (navbar) {
      navbar.dispose();
    }
  });

  it('should render a navbar with one element', function() {
    navbar = new ClayNavbar({
      items: [{title: 'Page 1', url: '#1'}],
    });

    expect(navbar).toMatchSnapshot();
  });

  it('should render an inverted colored navbar with one element', function() {
    navbar = new ClayNavbar({
      inverted: true,
      items: [{title: 'Page 1', url: '#1'}],
    });

    expect(navbar).toMatchSnapshot();
  });

  it('should render a navbar with one active element', function() {
    navbar = new ClayNavbar({
      items: [{active: true, title: 'Page 1', url: '#1'}],
    });

    expect(navbar).toMatchSnapshot();
  });

  it('should render a navbar with more than one element', function() {
    navbar = new ClayNavbar({
      items: [{title: 'Page 1', url: '#1'}, {title: 'Page 2', url: '#2'}],
    });

    expect(navbar).toMatchSnapshot();
  });

  it('should fail when no items are passed', function() {
    expect(() => {
      navbar = new ClayNavbar();
    }).toThrow();
  });

  it('should fail when no title is passed on each page', function() {
    expect(() => {
      navbar = new ClayNavbar({
        items: [{url: '#1'}],
      });
    }).toThrow();
  });

  it('should fail when no url is passed on each page', function() {
    expect(() => {
      navbar = new ClayNavbar({
        items: [{title: 'Page 1'}],
      });
    }).toThrow();
  });
});
