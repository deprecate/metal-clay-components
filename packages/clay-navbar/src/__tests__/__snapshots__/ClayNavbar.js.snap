// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayNavbar should render a navbar with more than one element 1`] = `
<nav class="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-light">
  <div class="container">
    <a aria-label="Toggle Navigation" aria-expanded="false" class=" navbar-toggler navbar-toggler-link collapsed" href="#">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
        <use xlink:href="icons.svg#caret-bottom"></use>
      </svg>
    </a>
    <div class="navbar-collapse collapse">
      <div class="container">
        <ul class="navbar-nav">
          <li class="nav-item">
            <a class=" nav-link" href="#1">
              <span class="navbar-text-truncate">Page 1</span>
            </a>
          </li>
          <li class="nav-item">
            <a class=" nav-link" href="#2">
              <span class="navbar-text-truncate">Page 2</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;

exports[`ClayNavbar should render a navbar with one active element 1`] = `
<nav class="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-light">
  <div class="container">
    <a aria-label="Toggle Navigation" aria-expanded="false" class=" navbar-toggler navbar-toggler-link collapsed" href="#">Page 1
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
        <use xlink:href="icons.svg#caret-bottom"></use>
      </svg>
    </a>
    <div class="navbar-collapse collapse">
      <div class="container">
        <ul class="navbar-nav">
          <li class="nav-item">
            <a aria-label="Current Page" class=" nav-link active" href="#1">
              <span class="navbar-text-truncate">Page 1</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;

exports[`ClayNavbar should render a navbar with one element 1`] = `
<nav class="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-light">
  <div class="container">
    <a aria-label="Toggle Navigation" aria-expanded="false" class=" navbar-toggler navbar-toggler-link collapsed" href="#">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
        <use xlink:href="icons.svg#caret-bottom"></use>
      </svg>
    </a>
    <div class="navbar-collapse collapse">
      <div class="container">
        <ul class="navbar-nav">
          <li class="nav-item">
            <a class=" nav-link" href="#1">
              <span class="navbar-text-truncate">Page 1</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;

exports[`ClayNavbar should render an inverted colored navbar with one element 1`] = `
<nav class="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-secondary">
  <div class="container">
    <a aria-label="Toggle Navigation" aria-expanded="false" class=" navbar-toggler navbar-toggler-link collapsed" href="#">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
        <use xlink:href="icons.svg#caret-bottom"></use>
      </svg>
    </a>
    <div class="navbar-collapse collapse">
      <div class="container">
        <ul class="navbar-nav">
          <li class="nav-item">
            <a class=" nav-link" href="#1">
              <span class="navbar-text-truncate">Page 1</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;
